br/pkg/restore/snap_client/pipeline_items.go

```diff
@@ -118,154 +116,269 @@ func (rc *SnapClient) RestorePipeline(ctx context.Context, plCtx PipelineContext
+						// ectx is cancelled, return and get the error from error group in defer function
+					case outCh <- cloneTable:
+					}
+					return nil
+				})
+			}
 		}
-	}
+	})
+	return outCh
 }
 
-// GoValidateChecksum forks a goroutine to validate checksum after restore.
-// it returns a channel fires a struct{} when all things get done.
-func (rc *SnapClient) GoValidateChecksum(
-	ctx context.Context,
-	inCh <-chan *CreatedTable,
+// registerValidateChecksum validates checksum after restore.
+func (rc *SnapClient) registerValidateChecksum(
+	builder *PipelineConcurrentBuilder,
 	kvClient kv.Client,
-	errCh chan<- error,
 	updateCh glue.Progress,
 	concurrency uint,
-) chan *CreatedTable {
-	log.Info("Start to validate checksum")
-	outCh := defaultOutputTableChan()
-	workers := tidbutil.NewWorkerPool(defaultChecksumConcurrency, "RestoreChecksum")
-	go concurrentHandleTablesCh(ctx, inCh, outCh, errCh, workers, func(c context.Context, tbl *CreatedTable) error {
+) {
+	builder.RegisterPipelineTask("Restore Checksum", defaultChecksumConcurrency, func(c context.Context, tbl *CreatedTable) error {
 		err := rc.execAndValidateChecksum(c, tbl, kvClient, concurrency)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		updateCh.Inc()
 		return nil
-	}, func() {
+	}, func(context.Context) error {
 		log.Info("all checksum ended")
+		return nil
 	})
-	return outCh
 }
 
-func (rc *SnapClient) GoUpdateMetaAndLoadStats(
-	ctx context.Context,
+const statsMetaItemBufferSize = 3000
+
+type statsMetaItemBuffer struct {
+	sync.Mutex
+	autoAnalyze bool
+	metaUpdates []statstypes.MetaUpdate
+}
+
+func NewStatsMetaItemBuffer(autoAnalyze bool) *statsMetaItemBuffer {
+	return &statsMetaItemBuffer{
+		autoAnalyze: autoAnalyze,
+		metaUpdates: make([]statstypes.MetaUpdate, 0, statsMetaItemBufferSize),
+	}
+}
+
+func (buffer *statsMetaItemBuffer) appendItem(item statstypes.MetaUpdate) (metaUpdates []statstypes.MetaUpdate) {
+	buffer.Lock()
+	defer buffer.Unlock()
+	buffer.metaUpdates = append(buffer.metaUpdates, item)
+	if len(buffer.metaUpdates) < statsMetaItemBufferSize {
+		return
+	}
+	metaUpdates = buffer.metaUpdates
+	buffer.metaUpdates = make([]statstypes.MetaUpdate, 0, statsMetaItemBufferSize)
+	return metaUpdates
+}
+
+func (buffer *statsMetaItemBuffer) take() (metaUpdates []statstypes.MetaUpdate) {
+	buffer.Lock()
+	defer buffer.Unlock()
+	metaUpdates = buffer.metaUpdates
+	buffer.metaUpdates = nil
+	return metaUpdates
+}
+
+func (buffer *statsMetaItemBuffer) UpdateMetasRest(ctx context.Context, statsHandler *handle.Handle) error {
+	metaUpdates := buffer.take()
+	if len(metaUpdates) == 0 {
+		return nil
+	}
+	return statsHandler.SaveMetaToStorage("br restore", false, metaUpdates...)
+}
+
+func (buffer *statsMetaItemBuffer) TryUpdateMetas(ctx context.Context, statsHandler *handle.Handle, physicalID, count int64) error {
+	item := statstypes.MetaUpdate{
+		PhysicalID:  physicalID,
+		Count:       count,
+		ModifyCount: 0,
+	}
+	if buffer.autoAnalyze {
```

Comment:
I think this flag should control the count as well. As we discussed yesterday, if the count is not 0, it will trigger the auto-analyze.

Reply:
Will be update by another PR